### ISSUE TITLE
Add an iterator (?)

### ISSUE THREAD

**Issue by BlueM** [BlueM/Tree#10]
Just an idea: would an iterator – to traverse the whole tree or part of the tree – be useful, in addition to the iteration methods currently available?

**Comment by PetengDedet** [comment]
+1
this feature would be very useful

**Comment by BlueM** [comment]
@PetengDedet It’s simple to implement `Iterator` in both `Tree` and `Tree\Node`, which would enable simple iteration, e.g. using `foreach`. But there are a number of possible ways what could be the desired behavior: flat iteration (rather not) or recursive iteration (probably preferred), and if recursive: depth-first or breadth-first.

I’d go for recursive and depth-first, but maybe you wouldn’t agree? For what purposes would you have had a need for iterating the tree?

**Comment by BlueM** [comment]
To see what it “feels” like, I made `Tree` and `Node` implement `Iterator` in branch https://github.com/BlueM/Tree/compare/v2-dev-iterator-spike, i.e., they can be used like this:

```php
foreach ($tree as $node) {
    // Do something with the node
}
```

This is nice and clean, but it doesn’t differ that much from what one can do currently:

```php
foreach ($tree->getNodes() as $node) {
    // Do something with the node
}
```

The outcome is exactly the same in both cases, except that the `Iterator`-based approach requires more code and in my tests, was usually 100% to 200% *slower* than using `$tree->getNodes()`.

Equally, using that branch, you can iterate over nodes, but again, the API is almost identical what is possible now (using `$node->getDescendants()`) – though I did not bother to benchmark the two approaches, but would expect a similar result.

So, to summarize the iterator approach:
* The amount of code is almost exactly the same (and the no. of LOC are 100% identical)
* More code
* Less speed
* Not more flexible. You could argue that it becomes more flexible when used with a `FilterIterator` (or other SPL outer iterators), but one can get the same result – probably with notably less code – by using ``array_filter()``.

Verdict: doesn’t make sense to me.

**Comment by BlueM** [comment]
Closing, as an internal iterator does not seem to make much sense and an external iterator does not necessarily have to be part of the library (or can be added later without BC break).
